File: tests/test_ladder.py
"""Tests for woodshed.ladder: the speed ladder's pure state machine.

Two counters matter here and must not be conflated (see CLAUDE.md and
ladder.py's module docstring): `clean_at_speed` is progress toward the next
rung (0..reps_to_advance, resets to 0 on advance); the ledger's total
historical cleans at a speed is a different number and belongs to
ledger.clean_by_speed, not to this module.
"""

from __future__ import annotations

from woodshed.ladder import LadderConfig, LadderState, on_clean, on_retract, rungs, starting_speed


def test_rungs_default_config_50_to_100_step_5() -> None:
    cfg = LadderConfig()
    assert rungs(cfg) == [50.0, 55.0, 60.0, 65.0, 70.0, 75.0, 80.0, 85.0, 90.0, 95.0, 100.0]


def test_rungs_includes_target_speed_exactly() -> None:
    cfg = LadderConfig(start_speed=90.0, ladder_step=5.0, target_speed=100.0)
    assert rungs(cfg) == [90.0, 95.0, 100.0]


def test_rungs_single_rung_when_start_equals_target() -> None:
    cfg = LadderConfig(start_speed=100.0, target_speed=100.0)
    assert rungs(cfg) == [100.0]


def test_remaining_is_reps_to_advance_minus_clean_at_speed() -> None:
    cfg = LadderConfig(reps_to_advance=3)
    state = LadderState(speed=50.0, clean_at_speed=1)
    assert state.remaining(cfg) == 2


def test_remaining_floors_at_zero() -> None:
    # clean_at_speed cannot exceed reps_to_advance in practice (on_clean advances
    # and resets before that happens), but remaining must never go negative if it
    # somehow did.
    cfg = LadderConfig(reps_to_advance=3)
    state = LadderState(speed=50.0, clean_at_speed=5)
    assert state.remaining(cfg) == 0


def test_hint_names_next_rung_and_reps_needed() -> None:
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3)
    state = LadderState(speed=55.0, clean_at_speed=2)
    assert state.hint(cfg) == "-> 60% after 1 more"


def test_hint_at_target_speed_names_no_further_rung() -> None:
    cfg = LadderConfig(target_speed=100.0)
    state = LadderState(speed=100.0, clean_at_speed=1)
    hint = state.hint(cfg)
    # No further rung exists; the hint must not claim one at 105%.
    assert "105" not in hint


def test_starting_speed_with_no_history_is_start_speed() -> None:
    cfg = LadderConfig(start_speed=50.0)
    assert starting_speed({}, cfg) == 50.0


def test_starting_speed_three_cleans_at_55_returns_60_not_55() -> None:
    # Named per the test contract: returning the earned rung itself would force
    # re-earning work already advanced past.
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3)
    clean_by_speed = {55.0: 3}
    assert starting_speed(clean_by_speed, cfg) == 60.0


def test_starting_speed_ignores_rung_with_insufficient_cleans() -> None:
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3)
    clean_by_speed = {55.0: 2}
    assert starting_speed(clean_by_speed, cfg) == 50.0


def test_starting_speed_uses_highest_qualifying_rung() -> None:
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3)
    clean_by_speed = {55.0: 3, 60.0: 3, 65.0: 1}
    assert starting_speed(clean_by_speed, cfg) == 65.0


def test_starting_speed_capped_at_target_speed() -> None:
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3, target_speed=100.0)
    clean_by_speed = {100.0: 3}
    assert starting_speed(clean_by_speed, cfg) == 100.0


def test_on_clean_increments_without_reaching_threshold() -> None:
    cfg = LadderConfig(reps_to_advance=3)
    state = LadderState(speed=50.0, clean_at_speed=0)
    next_state = on_clean(state, cfg)
    assert next_state.speed == 50.0
    assert next_state.clean_at_speed == 1


def test_on_clean_advances_rung_and_resets_counter_at_threshold() -> None:
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3)
    state = LadderState(speed=50.0, clean_at_speed=2)
    next_state = on_clean(state, cfg)
    assert next_state.speed == 55.0
    assert next_state.clean_at_speed == 0


def test_on_clean_never_advances_past_target_speed() -> None:
    # Named per the test contract: assert this explicitly starting one step below
    # target, so a threshold reached at the top rung does not overshoot.
    cfg = LadderConfig(start_speed=50.0, ladder_step=5.0, reps_to_advance=3, target_speed=100.0)
    state = LadderState(speed=100.0, clean_at_speed=2)
    next_state = on_clean(state, cfg)
    assert next_state.speed == 100.0
    assert next_state.clean_at_speed == 0


def test_on_retract_at_zero_leaves_speed_unchanged_and_does_not_go_negative() -> None:
    cfg = LadderConfig()
    state = LadderState(speed=55.0, clean_at_speed=0)
    next_state = on_retract(state, cfg)
    assert next_state.speed == 55.0
    assert next_state.clean_at_speed == 0


def test_on_retract_drops_by_one_not_to_zero() -> None:
    # Named for the punishment this avoids: one bad rep should not erase all
    # progress toward the rung.
    cfg = LadderConfig()
    state = LadderState(speed=55.0, clean_at_speed=2)
    next_state = on_retract(state, cfg)
    assert next_state.speed == 55.0
    assert next_state.clean_at_speed == 1


def test_on_retract_never_changes_speed() -> None:
    cfg = LadderConfig()
    state = LadderState(speed=55.0, clean_at_speed=1)
    next_state = on_retract(state, cfg)
    assert next_state.speed == 55.0

File: web/tests/test_ladder.mjs
// web/tests/test_ladder.mjs — K1's test contract for web/ladder.js, run as
// a plain node script (see test_seek.mjs for why this repo's front-end
// tests are not under a framework).
//
// The pure rules below are asserted here AND compared against
// src/woodshed/ladder.py itself by tests/test_ladder_mirror.py, which runs
// both over the same inputs. This file owns the cases that are about the
// BROWSER's half — auto-confirm, when an advance is allowed to happen —
// which the Python has no opinion about.

import assert from 'node:assert/strict';
import {
  Ladder,
  nextRung,
  onClean,
  onRetract,
  rungs,
  startingSpeed,
} from '../ladder.js';

let failures = 0;
function test(name, fn) {
  try { fn(); console.log(`ok - ${name}`); }
  catch (err) { failures += 1; console.error(`FAIL - ${name}`); console.error(err); }
}

const CFG = { startSpeed: 50, ladderStep: 5, repsToAdvance: 3, targetSpeed: 100 };

test('rungs run from the start speed to the target inclusive', () => {
  assert.deepEqual(rungs(CFG), [50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100]);
  assert.deepEqual(rungs({ ...CFG, ladderStep: 0 }), [50]);
});

test('nothing advances past the target speed', () => {
  assert.equal(nextRung(100, CFG), null);
  assert.deepEqual(onClean({ speed: 100, cleanAtSpeed: 2 }, CFG), { speed: 100, cleanAtSpeed: 0 });
});

test('starting speed is the rung ABOVE the one already earned', () => {
  // Returning 55 would make you re-earn work onClean already advanced past.
  assert.equal(startingSpeed({ 55: 3 }, CFG), 60);
  assert.equal(startingSpeed({}, CFG), 50);
  assert.equal(startingSpeed({ 55: 2 }, CFG), 50);
  assert.equal(startingSpeed({ 100: 9 }, CFG), 100);
});

test('a retraction drops progress by one, never to zero', () => {
  // The punishment it avoids: two good reps undone by one bad one.
  assert.deepEqual(onRetract({ speed: 55, cleanAtSpeed: 2 }, CFG), { speed: 55, cleanAtSpeed: 1 });
  assert.deepEqual(onRetract({ speed: 55, cleanAtSpeed: 0 }, CFG), { speed: 55, cleanAtSpeed: 0 });
});

// ---- the browser's half -------------------------------------------------

test('auto-confirm is on by default: a pass you played counts', () => {
  const ladder = new Ladder({ cfg: CFG });
  assert.equal(ladder.autoConfirm, true);
  const result = ladder.pass_();
  assert.equal(result.clean, true);
  assert.equal(ladder.cleanAtSpeed, 1);
});

test('with auto-confirm on, the human says "not that one" instead', () => {
  const ladder = new Ladder({ cfg: CFG });
  ladder.dirty();
  assert.equal(ladder.pass_().clean, false);
  assert.equal(ladder.cleanAtSpeed, 0);
  // And it only applies to the lap it was armed for.
  assert.equal(ladder.pass_().clean, true);
});

test('with auto-confirm off, only a confirmed lap counts', () => {
  const ladder = new Ladder({ cfg: CFG, autoConfirm: false });
  assert.equal(ladder.pass_().clean, false);
  ladder.confirm();
  assert.equal(ladder.pass_().clean, true);
  // A confirmation never carries over into the next lap.
  assert.equal(ladder.pass_().clean, false);
});

test('the advance happens on the pass, and fires once with both speeds', () => {
  const ladder = new Ladder({ cfg: CFG, speed: 55 });
  const events = [];
  ladder.addEventListener('advance', (e) => events.push(e.detail));
  ladder.pass_();
  ladder.pass_();
  assert.deepEqual(events, []);
  const third = ladder.pass_();
  assert.deepEqual(third.advanced, { from: 55, to: 60 });
  assert.equal(events.length, 1);
  assert.equal(events[0].to, 60);
  assert.equal(ladder.cleanAtSpeed, 0);
});

test('a rung is only ever advanced from pass_(), so never mid-loop', () => {
  // docs/03-audio-engine.md: "Speed changes at the boundary, never
  // mid-loop." Nothing else on this class can move `speed` upward.
  const ladder = new Ladder({ cfg: CFG });
  const before = ladder.speed;
  ladder.confirm();
  ladder.dirty();
  ladder.retract();
  assert.equal(ladder.speed, before);
});

test('a manual speed change resets progress toward the next rung', () => {
  // Otherwise two presses of speed_down and one clean rep would advance
  // you off a rung you never actually practised.
  const ladder = new Ladder({ cfg: CFG, speed: 60, cleanAtSpeed: 2 });
  ladder.setSpeed(50);
  assert.equal(ladder.cleanAtSpeed, 0);
  assert.equal(ladder.speed, 50);
});

test('retract undoes one rep of progress on the live ladder too', () => {
  const ladder = new Ladder({ cfg: CFG });
  ladder.pass_();
  ladder.pass_();
  assert.equal(ladder.cleanAtSpeed, 2);
  ladder.retract();
  assert.equal(ladder.cleanAtSpeed, 1);
  assert.equal(ladder.speed, 50);
});

test('hint names the next rung and how many more it wants', () => {
  assert.equal(new Ladder({ cfg: CFG, speed: 55 }).hint, '-> 60% after 3 more');
  assert.equal(new Ladder({ cfg: CFG, speed: 100 }).hint, '-> target after 3 more');
});

if (failures) process.exitCode = 1;
